lib/teiserver/battle/libs/balance_lib.ex

```diff
@@ -154,6 +155,30 @@ defmodule Teiserver.Battle.BalanceLib do
     |> Map.put(:time_taken, System.system_time(:microsecond) - start_time)
   end
 
+  @doc """
+  Sometimes groups have missing data so we need to refetch it.
+  If we go through balancer_server then all the required data should be there
+  """
+  def standardise_groups(groups) do
+    groups
+    |> Enum.map(fn group ->
+      # Iterate over our map
+      {_ignored, better_map} =
+        Enum.map_reduce(group, %{}, fn {user_id, value}, acc ->
+          fixed_value =
+            cond do
+              # We're missing data so need to fetch it
+              is_number(value) -> get_user_rating_rank_old(user_id, value)
+              true -> value
+            end
+
+          {nil, Map.put(acc, user_id, fixed_value)}
+        end)
+
+      better_map
+    end)
+  end
```

Comment:
If I understand this bit correctly, when the value in the map is not a user struct, you want to fetch the corresponding user and replace the id with that user right?

In that case there's a much simpler way: `Enum.map`

```elixir
    Enum.map(groups, fn group ->
      group
      |> Enum.map(fn {user_id, value} ->
        cond do
          is_number(value) -> {user_id, get_user_rating_rank_old(user_id, value)}
          true -> {user_id, value}
        end
      end)
      |> Enum.into(%{})
    end)
```

or, if you prefer using guards, you can also do the following

```elixir
    Enum.map(groups, fn group ->
      group
      |> Enum.map(fn
        {user_id, value} when is_number(value) ->
          {user_id, get_user_rating_rank_old(user_id, value)}

        {user_id, value} ->
          {user_id, value}
      end)
      |> Enum.into(%{})
    end)
```

The advantage of `Enum.map` is that the intent is clearer, you are applying some sort of transformation on every element. In the case of a map it's a tuple of `{key, value}`.
When using `map_reduce`, it's a lot more complicated to figure out what's going on. The fact that you had to discard one of the return value is a code smell as well.

Reply:
I had a feeling there was a better way to do it. Before you reviewed this, I had already posted a question on Elixir forums: https://elixirforum.com/t/iterate-through-map-of-dynamic-keys-and-update-their-values/64061/3
I'm liking the idea of Map.new